chromadb/rate_limiting/admission_controller.py

```diff
@@ -0,0 +1,358 @@
+import threading
+import time
+import concurrent.futures
+import queue
+
+from typing import Optional, cast
+
+
+class Ticket:
+    def __init__(self, issued_by: Optional["AdmissionController"]) -> None:
+        self.acquisition_elapsed = 0.0
+        self.issued_by = issued_by
+
+    def release(self) -> None:
+        if self.issued_by:
+            self.issued_by.release()
+            self.issued_by = None
+
+
+class Waiter:
+    def __init__(self) -> None:
+        self.index = 0
+        self.woke_up = False
+
+    def notify(self) -> None:
+        self.woke_up = True
+
+
+class AdmissionController:
+    def admit_one(self) -> Optional[Ticket]:
+        raise NotImplementedError
+
+    def release(self) -> None:
+        raise NotImplementedError
+
+    def admitted(self) -> int:
+        raise NotImplementedError
+
+    def capacity(self) -> int:
+        raise NotImplementedError
+
+    def resize(self) -> int:
+        raise NotImplementedError
+
+    def stop(self) -> None:
+        raise NotImplementedError
+
+
+class AdmissionControllerImpl(AdmissionController):
+    FIFO_MODE = 0
+    LIFO_MODE = 1
+    DEFAULT_M = 0.005
+    DEFAULT_N = 0.1
+
+    def __init__(self, parallelism: int, M: float = DEFAULT_M, N: float = DEFAULT_N):
+        self.mtx = threading.Lock()
+        self.stopped = False
+        self.M = M
+        self.N = N
+        self.allowed = parallelism
+        self.admitted_count = 0
+        self.last_empty = time.time()
+        self.queue_mode = self.FIFO_MODE
+        self.waiters: list[Optional[Waiter]] = [None] * 64
+        self.head = 0
+        self.tail = 0
+        self.waiter_condition = threading.Condition(self.mtx)
+
+    def admit_one(self) -> Optional[Ticket]:
+        admit = False
+        stopped = self.stopped
+
+        with self.mtx:
+            if self.head == self.tail and self.admitted_count < self.allowed:
+                self.admitted_count += 1
+                admit = True
+
+        if stopped:
+            return None
+
+        tk = Ticket(issued_by=self)
+        if admit:
+            return tk
+
+        return self.admit_one_slow_path(tk=tk)
+
+    def admit_one_slow_path(self, tk: Ticket) -> Optional[Ticket]:
+        t0 = time.time()
+        waiter = Waiter()
+        timeout, enqueued = self.enqueue_waiter(waiter=waiter)
+
+        if not enqueued:
```

Comment:
This looks like a bug.  We should fail out, not return a ticket.

Reply:
I understood this to mean that in the time between checking the queue in `admit_one` and in `enqueue_waiter` it emptied so we can admit on the fast path. That's what it seems like in the [Go implementation](https://github.com/dropbox/load_management/blob/de03a5c02a72c9e058e4075b00a2f81c36652175/admission_control/admission_control.go#L221). Am I missing something?

Reply:
Your understanding is correct.  I was mistaken.